src/Microsoft.Identity.Client/ApiEvent.cs

```diff
@@ -0,0 +1,48 @@
+﻿//----------------------------------------------------------------------
+//
+// Copyright (c) Microsoft Corporation.
+// All rights reserved.
+//
+// This code is licensed under the MIT License.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files(the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+//------------------------------------------------------------------------------
+
+
+namespace Microsoft.Identity.Client
+{
+    internal class ApiEvent : Event
+    {
+        public ApiEvent(string authority, string authorityType, string uiBehavior, string validationStatus, bool loginHint = false) : base("Microsoft.MSAL.api_event")
+        {
+            this["authority"] = authority;
+            this["authority_type"] = authorityType;
+            this["ui_behavior"] = uiBehavior;
+            this["validation_status"] = validationStatus;
+            this["login_hint"] = loginHint.ToString();
+        }
+
+        public void Update(string idToken, bool wasSuccessful)
```

Comment:
Agree. `Update` is not creative enough. :-)  I don't currently have a good name in mind, so let me describe its purpose here, then maybe our teammates will propose some names.

The event properties will be collected in 2 separated steps.

1. Right before the actual operation occurs, we already know all of its calling parameters. For example, before an HTTP request hits the wire, we already know the `http_path`, `query_parameters`, etc.. Creating an event with these information (and inserting it into the temporary container) gives us chance to know more (than only a `start_time`), in case an exception will happen later and the event won't be properly stopped.

2. After the operation finishes, MSAL can `Update()` the previous event with more information.

Actually the MSAL Android Telemetry PR implements similar behaviors, by using the builder pattern. You can use the builder to keep adding/updating data, and "convert" the builder instance into an event instance in the end. During the process, the builder instance is a de-facto partial event container. So it is the same.

The signature of constructor and the `Update()` method also regulates what can/should be provided in each scenario.

PS: Why your comment came in at 3am?